### ISSUE TITLE
Perl::Critic::Utils::all_perl_files option to skip symlinks

### ISSUE THREAD

**Issue by thaljef** [Perl-Critic/Perl-Critic#502]
RT Ticket: http://rt.cpan.org/Ticket/Display.html?id=83446
Requested On: Tue Feb 19 06:01:39 2013
Requested By: EDAVIS@cpan.org

---

**Comment by epa** [comment]
(pasting original bug report)

`all_perl_files('.')` will recurse into subdirectories, including symlinks. This can be slow if you have a symlink to some big directory. It would be useful to have an option to not follow symlinks.

Clearly the code change is simple, but a good interface for it is trickier. Currently `all_perl_files` takes a list of directory names; perhaps if the first argument is a code ref, it is used to filter directories:

````@files = all_perl_files sub { not(-l $_[0]) }, '.';````

Would you consider adding that enhancement to the library?

For the perlcritic command line tool, I'm tempted to say it should print a warning message such as 

````Looking inside symbolic-linked directories, such as ./foo````

if it finds such a thing and `--quiet` is not set. Then nobody would wonder why a perlcritic run that used to run quickly is now apparently hanging.

**Comment by thaljef** [comment]
I could imagine amending `all_perl_files` as you suggest, but only as a convenience. At some point, my answer is "use `File::Find`" (or equivalent) directly. I'm undecided about the warning message though.

**Comment by epa** [comment]
Yes, File::Find works; I just didn't want to reinvent the wheel with Perl::Critic's list of version control directories and backup files to skip, and its detection of what files look like Perl code.  There might even be a case for putting that code in its own tiny library released separately to CPAN.

For the warning on symlinks, there could be three settings `--symlinks=follow`, `--symlinks=nofollow`, and `--symlinks=warn` (the default).  These might also be configured in perlcriticrc.
